examples/distributed-spmv-scaling/distributed-spmv-scaling.cpp

```diff
@@ -0,0 +1,270 @@
+            for (IndexType d_j = -1; d_j <= 1; d_j++) {
+                for (IndexType d_k = -1; d_k <= 1; d_k++) {
+                    if (!restricted ||
+                        ((d_i == 0 && d_j == 0) || (d_i == 0 && d_k == 0) ||
+                         (d_j == 0 && d_k == 0))) {
+                        auto col =
+                            k + d_k + (j + d_j) * dp + (i + d_i) * dp * dp;
+                        if (col >= 0 && col < mat_size) {
+                            A_data.nonzeros.emplace_back(row, col,
+                                                         gko::one<ValueType>());
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return A_data;
+}
+
+
+int main(int argc, char* argv[])
+{
+    const auto fin = gko::mpi::init_finalize(argc, argv);
+    // Use some shortcuts. In Ginkgo, vectors are seen as a gko::matrix::Dense
+    // with one column/one row. The advantage of this concept is that using
+    // multiple vectors is a now a natural extension of adding columns/rows are
+    // necessary.
+    using ValueType = double;
+    using GlobalIndexType = gko::distributed::global_index_type;
+    using LocalIndexType = GlobalIndexType;
+    using dist_mtx = gko::distributed::Matrix<ValueType>;
+    using dist_vec = gko::distributed::Vector<ValueType>;
+    using vec = gko::matrix::Dense<ValueType>;
+    using part_type = gko::distributed::Partition<gko::int32>;
+
+    const auto comm = gko::mpi::communicator::create_world();
+    const auto rank = comm->rank();
+    const auto local_rank = comm->local_rank();
+
+    // Print the ginkgo version information.
+    if (rank == 0) {
+        std::cout << gko::version_info::get() << std::endl;
+    }
+
+    if (argc == 2 && (std::string(argv[1]) == "--help")) {
+        if (rank == 0) {
+            std::cerr << "Usage: " << argv[0]
+                      << " [executor] [DISCRETIZATION_POINTS] [2D] "
+                         "[RESTRICT_STENCIL] [STRONG_SCALING]"
+                      << std::endl;
+            std::cerr << "Default values:" << std::endl;
+            std::cerr << "      - executor:        reference" << std::endl;
+            std::cerr << "      - DISCRETIZATION_POINTS: 100" << std::endl;
+            std::cerr << "      - 2D:                      1" << std::endl;
+            std::cerr << "      - RESTRICT_STENCIL:        0" << std::endl;
+            std::cerr << "      - STRONG_SCALING:          1" << std::endl;
+        }
+        std::exit(-1);
+    }
+
+    const auto executor_string = argc >= 2 ? argv[1] : "reference";
+    std::map<std::string, std::function<std::shared_ptr<gko::Executor>()>>
+        exec_map{
+            {"omp", [] { return gko::OmpExecutor::create(); }},
+            {"cuda",
+             [local_rank] {
+                 return gko::CudaExecutor::create(
+                     local_rank, gko::ReferenceExecutor::create(), true);
```

Comment:
Continuing along @tcojean 's questions, right now this might not work when there's only device 0 on each process, right? Even if the device ID is global and independent of which process is assigned what resource by the scheduler, there's no guarantee device ID 3 would be accessible to rank 3.

Would it not be better to either (1) always assume device 0, assuming the user maps one GPU to one process in the job scheduler (`--gpus-per-task=1` on slurm, I think), or (2) use hwloc or something to access the number of GPUs available to this task and then decide `device_id`. Option 2 becomes useful only when we want to support multiple GPUs per process, so maybe we ignore that for now.

Reply:
I'm not sure what is best to take care of that right now, but I think that's a fairly important point and a fairly annoying one. AFAIK, both situations can happen where only one device (number 0) is available for each MPI process (as you say using some job scheduler option), or each process sees the full machine and thus GPUs need to be split like now.

We probably should work in both situations? But what would be a convenient way? Should we have a simple function somewhere doing `max(available_devices, local_rank)`?

Reply:
According to [this](https://docs.olcf.ornl.gov/systems/summit_user_guide.html#gpu-compute-modes), on Summit, the default setting is to have one GPU per process. On Spock, the default setting is all GPUs see all processes. I guess we need to support both, or document strongly that we need users to set the correct scheduler settings.

Reply:
I think that is regarding assignment rather than what the process can view. With Summit, AFAIK, it is a bit different with resource sets. I think you always need to specify how many GPUs you need. And the numbering depends on how many resource sets you ask for. By default, I think it is set to 1, so the GPUs will be numbered from 0 to 5.